source/magpie/operation.d

```diff
@@ -0,0 +1,337 @@
+module magpie.operation;
+
+/*
+ * This file will contain work on operations of aggregate and join
+ * aggregate:   Operation to work on a row/ column of DataFrame to give results like mean/median/mode/elc
+ * merge:       It joins two DataFrame in a way similar to that in Pandas or any RDBMS software
+ */
+
+import magpie.dataframe: DataFrame;
+
+/++
+Description: Merges two DataFrames based on indexes
+@params: type - Type of join: Inner, Outer, Left, Right
+@params: df1 - The left DataFrame
+@params: df2 - The right DataFrame
+@params: lsuffix - Suffix to add in case there is an index collision between left and right DataFrame
+@parmas: rsuffix - Suffix to add in case there is an index collision between left and right DataFrame
++/
+auto merge(string type = "", T, U)(T df1, U df2, string lsuffix = "x_", string rsuffix = "y_")
+{
+
+    assert(df1.indx.row.codes.length == df2.indx.row.codes.length,
+        "Index level mismatch. Cannot merge dataframes with different level of indexing");
+
+    if(df1.indx.column.codes.length > df2.indx.column.codes.length)
+    {
+        int[] codes;
+        codes.length = df2.indx.column.codes[0].length;
+
+        foreach(i; 0 .. df1.indx.column.codes.length - df2.indx.column.codes.length)
+        {
+            df2.indx.column.index ~= [[""]];
+            df2.indx.column.codes ~= [codes];
+        }
+    }
+    else if(df1.indx.column.codes.length < df2.indx.column.codes.length)
+    {
+        int[] codes;
+        codes.length = df1.indx.column.codes[0].length;
+        
+        foreach(i; 0 .. df2.indx.column.codes.length - df1.indx.column.codes.length)
+        {
+            df1.indx.column.index ~= [[""]];
+            df1.indx.column.codes ~= [codes];
+        }
+    }
+
+    DataFrame!(true, df1.RowType, df2.RowType) combinator;
+
+    import magpie.helper: ensureUnique, transposed;
+    combinator.indx.column.index = transposed(ensureUnique!1(df1.indx, df2.indx, lsuffix, rsuffix));
+
+    combinator.indx.generateCodes();
+    combinator.indx.optimize();
+
+    string[] unpack(size_t i)
+    {
+        string[] indx;
+        indx.length = combinator.indx.row.codes.length;
+        foreach(j; 0 .. combinator.indx.row.codes.length)
+        {
+            if(combinator.indx.row.index[j].length == 0)
+            {
+                import std.conv: to;
+                indx[j] = to!string(combinator.indx.row.codes[j][i]);
+            }
+            else
+            {
+                indx[j] = combinator.indx.row.index[j][combinator.indx.row.codes[j][i]];
+            }
+        }
+
+        return indx;
+    }
+
+    static if(type == "left")
```

Comment:
You should create a set operation enum for all these types of operations and have that as the type of `type` or add a constraint in place of the static assert at the bottom

Reply:
Addressed in 6517b4a